File: src/lib/tauri-api/pr-draft.ts
import type { WorkspacePrDraft, WorkspacePrResult, WorkspacePrStatus } from '../../types/pr-draft';
import { invokeCommand } from './client';

const PR_STATUS_CACHE_PREFIX = 'forge:workspace-pr-status:';

function prStatusCacheKey(workspaceId: string): string {
  return `${PR_STATUS_CACHE_PREFIX}${workspaceId}`;
}

export function getCachedWorkspacePrStatus(workspaceId: string): WorkspacePrStatus | null {
  if (typeof window === 'undefined') return null;
  try {
    const raw = window.localStorage.getItem(prStatusCacheKey(workspaceId));
    return raw ? JSON.parse(raw) as WorkspacePrStatus : null;
  } catch {
    return null;
  }
}

function cacheWorkspacePrStatus(workspaceId: string, status: WorkspacePrStatus): void {
  if (typeof window === 'undefined') return;
  try {
    window.localStorage.setItem(prStatusCacheKey(workspaceId), JSON.stringify(status));
  } catch {
    // Best-effort speed cache only.
  }
}

export function getWorkspacePrDraft(workspaceId: string): Promise<WorkspacePrDraft> {
  return invokeCommand<WorkspacePrDraft>('get_workspace_pr_draft', { workspaceId });
}

export function refreshWorkspacePrDraft(workspaceId: string): Promise<WorkspacePrDraft> {
  return invokeCommand<WorkspacePrDraft>('refresh_workspace_pr_draft', { workspaceId });
}

export function createWorkspacePr(workspaceId: string): Promise<WorkspacePrResult> {
  return invokeCommand<WorkspacePrResult>('create_workspace_pr', { workspaceId });
}

export async function getWorkspacePrStatus(workspaceId: string): Promise<WorkspacePrStatus> {
  const status = await invokeCommand<WorkspacePrStatus>('get_workspace_pr_status', { workspaceId });
  cacheWorkspacePrStatus(workspaceId, status);
  return status;
}

File: src-tauri/src/commands/pr_draft.rs
use tauri::State;

use crate::models::{WorkspacePrDraft, WorkspacePrResult, WorkspacePrStatus};
use crate::services::pr_draft_service;
use crate::state::AppState;

#[tauri::command]
pub fn get_workspace_pr_draft(
    state: State<'_, AppState>,
    workspace_id: String,
) -> Result<WorkspacePrDraft, String> {
    pr_draft_service::get_workspace_pr_draft(&state, &workspace_id)
}

#[tauri::command]
pub fn refresh_workspace_pr_draft(
    state: State<'_, AppState>,
    workspace_id: String,
) -> Result<WorkspacePrDraft, String> {
    pr_draft_service::refresh_workspace_pr_draft(&state, &workspace_id)
}

#[tauri::command]
pub fn create_workspace_pr(
    state: State<'_, AppState>,
    workspace_id: String,
) -> Result<WorkspacePrResult, String> {
    pr_draft_service::create_workspace_pr(&state, &workspace_id)
}

#[tauri::command]
pub fn get_workspace_pr_status(
    state: State<'_, AppState>,
    workspace_id: String,
) -> Result<WorkspacePrStatus, String> {
    pr_draft_service::get_workspace_pr_status(&state, &workspace_id)
}
